cibyl/sources/zuuld/source.py

```diff
@@ -0,0 +1,38 @@
+"""
+#    Copyright 2022 Red Hat
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+
+from cibyl.sources.source import Source, speed_index
+from cibyl.sources.zuuld.git_api import ZuulGitHub, ZuulLocal
+
+
+class ZuulD(Source):
+    def __init__(self, name: str, driver: str, **kwargs):
+        super().__init__(name, driver, **kwargs)
+        self.repos = kwargs.get("repos")
+        self.kwargs = kwargs
+        remote = kwargs.get("remote", False)
+        if not remote:
+            self.api = ZuulLocal(self.kwargs)
+        else:
+            self.api = ZuulGitHub(self.kwargs)
+        self.api._validate()
```

Comment:
For me, the only change that it needs before merging is moving this line into a `setup` method so that the cloning can be done only if the source can actually be used ([here it is called](https://github.com/rhos-infra/cibyl/blob/61a8ac3bcefa048799023330c704fc5ab5e3a6a5/cibyl/orchestrator.py#L103))

Reply:
I'm calling setup method from `_validate()` [Link](https://github.com/rhos-infra/cibyl/pull/182/files/d7a95e919c91e205f877accb33410953956bc122#diff-6e4fc470028a04815e6b323edf998122447fca3beced742e58e2bd190895cf7aR143), If i give it call from setup then it will go in recursion.

Reply:
Well i'm totally agree with you.. if you want i can fix that in followup patch.

Reply:
In _validate it calls the setup method of GitSource, here we need a setup method for the ZuulD source, that can be called from outside to delay cloning/pulling the repos

Reply:
Done.